python/ml4ir/base/model/architectures/auto_dag_network.py

```diff
@@ -0,0 +1,355 @@
+from __future__ import annotations
+
+from typing import List, Dict, Union, Type
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+from ml4ir.base.config.keys import FeatureTypeKey
+from ml4ir.base.features.feature_config import FeatureConfig
+from ml4ir.base.io.file_io import FileIO
+
+
+def get_layer_subclasses() -> Dict[str, Type[layers.Layer]]:
+    """Get mapping of {subclass-name: subclass} for all derivative classes of keras.layers.Layer"""
+
+    def full_class_name(cls):
+        """Get {package_name}.{class_name}"""
+        module = cls.__module__
+        if module == "builtins":
+            return cls.__qualname__  # avoid outputs like 'builtins.str'
+        return f"{module}.{cls.__qualname__}"
+
+    def get_sub_classes(cls):
+        """DFS traversal to get all subclasses of `cls` class"""
+        for subcls in cls.__subclasses__():
+            if subcls not in subclasses:
+                # Handles duplicates
+                subclasses[full_class_name(subcls)] = subcls
+                get_sub_classes(subcls)
+
+    subclasses = {}
+    get_sub_classes(layers.Layer)
+    return subclasses
+
+
+class CycleFoundException(Exception):
+    """Thrown when a cycle is identified in a DAG"""
+    pass
+
+
+class LayerNode:
+    """Defines a node in the `LayerGraph`"""
+
+    def __init__(
+            self,
+            name: str,
+            layer: layers.Layer = None,
+            dependent_children: List[LayerNode] = None,
+            inputs: List[str] = None,
+            inputs_as_list: bool = False
+    ):
+        """
+        Constructor to create a node
+
+        Parameters
+        ----------
+        name: str
+            Name of the node
+        layer: instance of type tf.keras.layers.Layer
+            Node operation, i.e. Layer for forward prop
+        dependent_children: list of LayerNode
+            All downstream nodes which depend on this node's output
+        inputs: list of str
+            All inputs to the `layer` to be passed during forward prop
+        inputs_as_list: bool
+            Boolean to indicate if the set of input tensors to the layer should be converted to list
+        """
+        self.name = name
+        self.layer = layer
+        self.dependent_children = dependent_children if dependent_children is not None else []
+        self.inputs = inputs
+        self.is_input_node = True if inputs is None else False
```

Comment:
[Minor] @arvindsrikantan `is_input_node` is probably confusing terminology given that you also have `inputs` in this same class. With that said, I don't have ideas for a better name.

Reply:
But my point is "input" means two different things in this class.

Reply:
ping